packages/core/src/lib/auth0-synchronization-service.ts

```diff
@@ -0,0 +1,74 @@
+import { User, UserWrite } from "@dotkomonline/types"
+import { UserService } from "../modules/user/user-service"
+import { Auth0Repository } from "./auth0-repository"
+
+// Id token returned from Auth0. We don't want core to depend on next-auth, so we duplicate the type here.
+type JWTToken = {
+  email?: string | null
+  sub?: string | null
+  name?: string | null
+  givenName?: string | null
+  familyName?: string | null
+}
+
+/**
+ * Synchronize users in a local database user table with Auth0.
+ */
+export interface Auth0SynchronizationService {
+  /**
+   * If no record of the user exists in the local database, save it to the database.
+   */
+  createUser: (token: JWTToken) => Promise<User>
+
+  /**
+   * Synchronize record of user in database with user data from Auth0.
+   */
+  synchronizeUser: (user: User) => Promise<User | undefined>
+}
+
+export class Auth0SynchronizationServiceImpl implements Auth0SynchronizationService {
+  constructor(
+    private readonly userService: UserService,
+    private readonly auth0Repository: Auth0Repository
+  ) {}
+
+  async createUser(token: JWTToken) {
+    if (!token.email || !token.sub || !token.name || !token.givenName || !token.familyName) {
+      throw new Error("Missing user data in claims")
+    }
+    const userData: UserWrite = {
+      auth0Sub: token.sub,
+      studyYear: -1,
+      email: token.email,
+      name: token.name,
+      givenName: token.givenName,
+      familyName: token.familyName,
+      lastSyncedAt: new Date(),
+    }
+
+    return this.userService.createUser(userData)
+  }
+
+  // if user.updatedAt is more than 1 day ago, update user
+  async synchronizeUser(user: User) {
+    const oneDay = 1000 * 60 * 60 * 24
+    const oneDayAgo = new Date(Date.now() - oneDay)
+    if (!user.lastSyncedAt || user.lastSyncedAt < oneDayAgo) {
+      console.log("updating user", user.id, user.auth0Sub)
+      const idpUser = await this.auth0Repository.getBySubject(user.auth0Sub)
+
+      if (idpUser === undefined) {
+        throw new Error("User does not exist in Auth0")
+      }
+
+      return this.userService.updateUser(user.id, {
+        email: idpUser.email,
+        givenName: idpUser.givenName,
+        familyName: idpUser.familyName,
+        name: `${idpUser.givenName} ${idpUser.familyName}`,
+        lastSyncedAt: new Date(),
+      })
+    }
+    return undefined
+  }
+}
```

Comment:
I would rather think you just perform a sync whenever you get an `id_token` (which is every time a user has to log in I think, and maybe whenever you refresh the access token? I do not know how much nextAuth does though), and that whenever the user wants to change something that we store in auth0, what is stored in auth0 is changed first, then we invalidate the `id_token` and fetch a new one? 

The 1 day is arbitrary and `id_token`s has it's own expiry (which by default is 10 hours). 

This kind of syncing just leads to a one-day gap between changing e.g. givenName and it actually being updated, if we forget to invalidate the id_token?

Reply:
> I would rather think you just perform a sync whenever you get an id_token (which is every time a user has to log in I think, and maybe whenever you refresh the access token? I do not know how much nextAuth does though), and that whenever the user wants to change something that we store in auth0, what is stored in auth0 is changed first, then we invalidate the id_token and fetch a new one?

I created this issue for it: https://linear.app/dotkom/issue/DOT-629/find-a-better-solution-for-synchronizing-auth0-user-data-to-ow-user. I'm not quite sure how this is done and I feel blocked by this PR on my #804 PR. In my opinion this is a case of optimization in regards to the amount of database calls (we fetch the user from the database on every `session` callback). I don't see what synchronization problems it leads to. So I would rather just merge this now and find a more optimized solution later.

Thoughts?

> This kind of syncing just leads to a one-day gap between changing e.g. givenName and it actually being updated, if we forget to invalidate the id_token?

When updating givenName we will use the auth0 mgmt api, then sync the user with the database. We would also do it this way if we did not store any user info in the application database. In other words, we will have the same problem with invalidating tokens no matter what right? I don't see how this is relevant to this synchronization approach?

Reply:
From what I can tell, a potential syncing problem would exist in either case?

I agree that it's probably easiest to merge this now, and worry about moving to synchronizing against id_token in a future PR